File: packages/dotnet/src/Yourdudeken.Mpesa/Services/B2BService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Yourdudeken.Mpesa.Config;
using Yourdudeken.Mpesa.Http;
using Yourdudeken.Mpesa.Auth;

namespace Yourdudeken.Mpesa.Services
{
    public class B2BService
    {
        private readonly HttpClientWrapper _httpClient;
        private readonly AuthService _auth;
        private readonly MpesaConfig _config;
        private readonly string _baseUrl;

        public B2BService(HttpClientWrapper httpClient, AuthService auth, MpesaConfig config)
        {
            _httpClient = httpClient;
            _auth = auth;
            _config = config;
            _baseUrl = httpClient.GetBaseUrl();
        }

        public async Task<string> Send(string receiverShortcode, string commandId, int amount, string remarks,
            string accountNumber = null, string resultUrl = null, string timeoutUrl = null, string shortCodeType = "B2B")
        {
            if (commandId == "BusinessPayBill" && string.IsNullOrEmpty(accountNumber))
                throw new ArgumentException("Account Number is required for BusinessPayBill CommandID");

            var url = $"{_baseUrl}/mpesa/b2b/v1/paymentrequest";
            var body = new Dictionary<string, object>
            {
                { "Initiator", _config.InitiatorName },
                { "SecurityCredential", GenerateSecurityCredential() },
                { "CommandID", commandId },
                { "SenderIdentifierType", "4" },
                { "RecieverIdentifierType", "4" },
                { "Amount", amount },
                { "PartyA", _config.B2cShortcode },
                { "PartyB", receiverShortcode },
                { "AccountReference", accountNumber },
                { "Remarks", remarks },
                { "ResultURL", resultUrl ?? _config.B2bResultUrl },
                { "QueueTimeOutURL", timeoutUrl ?? _config.B2bTimeoutUrl }
            };

            var token = await _auth.GetAccessToken(shortCodeType);
            return await _httpClient.Post(url, body, token);
        }

        private string GenerateSecurityCredential() => "";
    }
}

File: packages/java/src/main/java/com/yourdudeken/mpesa/services/B2BService.java
package com.yourdudeken.mpesa.services;

import com.yourdudeken.mpesa.config.MpesaConfig;
import com.yourdudeken.mpesa.http.HttpClient;
import com.yourdudeken.mpesa.auth.AuthService;
import java.util.HashMap;
import java.util.Map;

public class B2BService {
    private HttpClient httpClient;
    private AuthService auth;
    private MpesaConfig config;
    private String baseUrl;

    public B2BService(HttpClient httpClient, AuthService auth, MpesaConfig config) {
        this.httpClient = httpClient;
        this.auth = auth;
        this.config = config;
        this.baseUrl = httpClient.getBaseUrl();
    }

    public String send(String receiverShortcode, String commandId, int amount, String remarks,
                      String accountNumber, String resultUrl, String timeoutUrl, String shortCodeType) {
        if (commandId.equals("BusinessPayBill") && (accountNumber == null || accountNumber.isEmpty())) {
            throw new IllegalArgumentException("Account Number is required for BusinessPayBill CommandID");
        }

        String url = baseUrl + "/mpesa/b2b/v1/paymentrequest";
        Map<String, Object> body = new HashMap<>();
        body.put("Initiator", config.getInitiatorName());
        body.put("SecurityCredential", generateSecurityCredential());
        body.put("CommandID", commandId);
        body.put("SenderIdentifierType", "4");
        body.put("RecieverIdentifierType", "4");
        body.put("Amount", amount);
        body.put("PartyA", config.getB2cShortcode());
        body.put("PartyB", receiverShortcode);
        body.put("AccountReference", accountNumber);
        body.put("Remarks", remarks);
        body.put("ResultURL", resultUrl != null ? resultUrl : config.getB2bResultUrl());
        body.put("QueueTimeOutURL", timeoutUrl != null ? timeoutUrl : config.getB2bTimeoutUrl());

        String token = auth.getAccessToken(shortCodeType);
        return httpClient.post(url, body, token);
    }

    private String generateSecurityCredential() {
        return "";
    }
}
